Add SectionTitle; Add utility for focusing using arrow keys

rpominov opened the pull request:
## Description

Closes #1100
Closes #1271

1. Added `SectionTitle` https://section-title--638245578b47d4399fba191a.chromatic.com/?path=/story/library-section-title--demo
2. Added basic `ArrowFocus` utility
3. Used the new `SectionTitle` in `CollapsibleSection`, so props and style panels now use the new `SectionTitle`
4. Made `CollapsibleSection` open by default as specified in #1100

## Steps for reproduction

1. Open style panel
2. Open props panel
3. Check storybook demos by the links above

## Code Review

- [ ] hi @kof, I need you to do
  - conceptual review (architecture, feature-correctness)
- [ ] hi @istarkov, I need you to
  - review the `arrow-focus` utility to see whether it'll work for you in backgrounds 
- [ ] hi @taylornowotny, I need you to
  - check whether `SectionTitle` component looks and works as expected by the link above

## Before requesting a review

- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio-builder/blob/main/apps/builder/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env.example` and the `builder/env-check.js` if mandatory

kof commented:
- [ ] Input With Nested Buttons

this one is def. broken because arrows now don't allow to move caret in the input

kof commented on packages/design-system/src/components/primitives/arrow-focus.stories.tsx:
```diff
@@ -0,0 +1,108 @@
+  setArrowFocusColumn,
+} from "./arrow-focus";
+
+const main = {
```
I used to use this style initially but then realize if we don't wrap it into css() and don't use CSS type , you don't get the validation by ts

rpominov replied:
This is stories so I didn't think to much about it. Does it matter in stories?

kof replied:
yeah, idk

kof commented on packages/design-system/src/components/primitives/arrow-focus.tsx:
```diff
@@ -0,0 +1,162 @@
+import type { KeyboardEvent } from "react";
+
+const focusableSelector = (extra: string) =>
```
I am surprised you ended up not using react-aria, I thought we decided to use it?

rpominov replied:
It doesn't do what we need: https://discord.com/channels/955905230107738152/973300187567095858/1085820876060106812

rpominov replied:
Talked in DM: I misunderstood what FocusScope is, need to have another look. Maybe we can use it.

kof commented on packages/design-system/src/components/primitives/arrow-focus.tsx:
```diff
@@ -0,0 +1,162 @@
+  ArrowRight: ["horizontal", "+"],
+};
+
+const ROW_ATTRIBUTE = "data-arrow-focus-row";
```
why isit not using role="row" attribute like react aria does?

kof commented on packages/design-system/src/components/section-title.tsx:
```diff
@@ -0,0 +1,174 @@
+  },
+});
+
+const buttonStyle = css({
```
hmm? isn't this supposed to be our button component?

rpominov replied:
This is not our button. This is just styles for `<button>` to make it take entire space of the title. This is what you click when you click on the title.

kof replied:
damn, ok lets call it titleButtonStyle, because I totally thought its the + button there

kof commented on packages/design-system/src/components/section-title.tsx:
```diff
@@ -0,0 +1,174 @@
+    border: `2px solid ${theme.colors.borderFocus}`,
+  },
+  variants: {
+    hasAddIcon: { true: { paddingRight: theme.spacing[16] } },
```
hasAddButton?

istarkov commented on packages/design-system/src/components/primitives/arrow-focus.stories.tsx:
```diff
@@ -0,0 +1,108 @@
+            <li key={row}>
+              {["1", "2", "3"].map((column) => (
+                <button
+                  {...setArrowFocusRow(row)}
```
setArrowFocusRow looks like a not usual naming for things like this.
For example react-aria names are `focusWithinProps`, `hoverProps`
Downshift has namings like: `getLabelProps`, `getMenuProps` etc. 

For me would be good to have a name like `arrowFocusRowProps` or `getArrowFocusRowProps`

istarkov commented on packages/design-system/src/components/primitives/arrow-focus.tsx:
```diff
@@ -0,0 +1,162 @@
+    "[contenteditable]",
+    "[tabindex]",
+  ]
+    .map((selector) => `${selector}:not([disabled]):not([hidden])${extra}`)
```
Never seen `hidden` attribute in a wild.

Also instead of `:not[disabled])` you can just use  
```
  'button:enabled',
  'select:enabled',
  'textarea:enabled',
  'input:enabled',
```
in your list

TrySound replied:
hidden is basically html version of display: none. Allows to not rebuild dom.

TrySound replied:
Sick, never knew there is :enabled

istarkov commented on packages/design-system/src/components/primitives/arrow-focus.tsx:
```diff
@@ -0,0 +1,162 @@
+  | "first-per-row"
+  | "first-per-column";
+
+const getFocusable = (
+  filter: Filter,
+  root: Element,
+  current: Element
```
The assumption that current is the currently focused element?

```
root = event.currentTarget, 
current = event.target
```

istarkov commented on packages/design-system/src/components/primitives/arrow-focus.tsx:
```diff
@@ -0,0 +1,162 @@
+    if (filter !== "all") {
+      const attribute =
+        filter === "within-row" ? ROW_ATTRIBUTE : COLUMN_ATTRIBUTE;
+      extraSelector = `[${attribute}="${current.getAttribute(attribute)}"]`;
```
Need to be commented on why we want the attribute to have the same value as current[attribute]?

rpominov commented:
@istarkov I'll most likely remove the custom utility and replace it with react-aria's FocusScope after all

istarkov commented on packages/design-system/src/components/section-title.stories.tsx:
```diff
@@ -0,0 +1,85 @@
+          isOpen={isOpen}
+          onOpenChange={setOpen}
+          hasItems={["local", "remote"]}
+          {...props}
```
please don't use spreads especially in examples 
`onAdd` is it come from props or how?

rpominov replied:
I don't see anything wrong with spread here. 

Intent here: `Wrapped` wraps `SectionTitle` with some code that otherwise would have been repeated a lot in demos, it controls some of the props, and passes through the rest. 

Spread is perfect to express: "passes through the rest".

rpominov replied:
Talked in DM: I didn't know something like this is not a TS error:

```
const props = {invalidProp: "something"}
<Something foo="bar" {...props} />
```

I agree it's a good idea to avoid props spread when possible because of this.

istarkov commented on packages/design-system/src/components/section-title.tsx:
```diff
@@ -0,0 +1,174 @@
+            </div>
+          )}
+        </button>
+        {onAdd && (
```
I always don't like logic like this, if we have a handler then we draw UI element.
I would prefer to have here only placeholder for the Add button like its done https://github.com/webstudio-is/webstudio-builder/blob/43617f1a59aa2eb4f9a385ca08b778f69b5f8f24/packages/design-system/src/components/css-value-list-item.tsx#L112

istarkov replied:
The issue is opening logic. 

Also its interesting how to use that `onAdd` in our UI having that we render Section Collapsible elements and style Section in different places. 
How to call onAdd inside Background section if it called on some parent

rpominov replied:
> The issue is opening logic.

Yeah, main reason I've put button inside the component instead of having a `suffix` prop or something is because I wanted to pass the `onClick` callback from the component. And there're also some minor reasons.

> Also its interesting how to use that onAdd in our UI having that we render Section Collapsible elements and style Section in different places.
How to call onAdd inside Background section if it called on some parent
Or we can make SectionTitle components as like as we do custom Sections.

Yeah, we'll have to refactor sections a bit somehow. One options is to make each section render the entire collapsible instead of just its content.

istarkov reviewed (approved):
Super!

taylornowotny commented:
Two issues with the component:

   1.  The "open" state of the section should have black icons (token: foreground/icon/main)

  2.  What is the "open, but empty" variant for? When closed it shows a blue dot, which would indicate that it is not empty.

rpominov commented:
> What is the "open, but empty" variant for? 

This is how the following logic is implemented in code:

> Sections that do not have a an Add (+) button are open by default. Sections that do have the button are closed by default. The + button adds items to the section and each item will have a - button to remove them.

In code:
 - **All** sections are open by default
 - If sections has items that can be added/removed and currently empty, then it looks like closed even if it's open

> When closed it shows a blue dot, which would indicate that it is not empty.

In the demo, when you click on it an item gets added, so it's no longer empty after that

rpominov commented:
> The "open" state of the section should have black icons (token: foreground/icon/main)

The + icon is black in the open state currently. Maybe you've looked at "open, but empty"

rpominov commented:
- [x] Fixed minor review comments such as renaming or props spread
- [x] Replaced custom focus manager with FocusScope (removed rows/columns support for now, don't want to do it as part of this PR). This resolves all comments on the custom utility that is removed now
- [x] Added Label as specified in: https://github.com/webstudio-is/webstudio-builder/issues/1271
- [ ] There's still an open question regarding how click/focus should be handled on the Label https://github.com/webstudio-is/webstudio-builder/issues/1271#issuecomment-1477611492

rpominov commented on packages/design-system/src/components/primitives/arrow-focus.tsx:
```diff
@@ -0,0 +1,31 @@
+import { FocusScope, useFocusManager } from "react-aria";
```
Actually they seem to allow to install individual packages https://react-spectrum.adobe.com/react-aria/getting-started.html#installation

Will install it
